### ISSUE TITLE
Handle re-exports

### ISSUE THREAD

**Issue by remcohaszing** [lydell/eslint-plugin-simple-import-sort#21]
Some modules re-export members directly. This is also sort of an import.

I.e.

```js
export { default as Foo, Fooz, Foob } from './Foo';
export { default, NonDefault } from './qwe';
export * from './asd';
export * as bar from './bar';
```

It would be nice if this would be sorted automatically as well following the same rules as imports.

I think this is mostly not combined with regular imports, but it could be. This means there should be rules for this.

Candidate rules I could come up with:

1. re-exports before imports
2. re-exports after imports
3. re-exports mixed with imports

Personally I’d go with 2.

**Comment by lydell** [comment]
Technically those are exports, but I agree that they feel like imports.

**Comment by SpainTrain** [comment]
Does the maybe tag indicate willingness to accept a PR?

When using this plugin, one begins to expect any statement using `from` syntax to sort by the value after `from`.  We've been tripped up by this a few times.  I think consistent use of a particular syntax (to find an item in a list of statements using that syntax) is more relevant than the different semantics between import and export.  Thoughts?

**Comment by SpainTrain** [comment]
Great plugin, BTW!

**Comment by lydell** [comment]
> Does the maybe tag indicate willingness to accept a PR?

Yes!

But keep in mind that it might not be merged. Even if I started working on this myself I might end up throwing it away depending on what I learned along the way.

The reason I haven't done this is because I very rarely use `export ... from` myself and as such haven’t felt like doing the hard work of figuring out exactly how this should work. I’ve spent _days_ just _thinking_ about how imports should work in all cases.

So feel free to start experimenting with a pull request if you want. I think it would be good to show a proof of concept early on to avoid perfecting something in vain.

But I fear that you’ll make a really nice pull request and I’ll get stuck reviewing it because I’d be too lazy to really wrap my head around it all and verify that we’ve covered all use cases and edge cases.

> (Great plugin BTW!)

Thanks! 😄

**Comment by JCrepin** [comment]
Hi all, I've created a POC PR for this. Let me know what you think.

#43

**Comment by lydell** [comment]
@remcohaszing I’m finally circling back to this. The reason I never merged #53 was because I didn’t feel 100% satisfied with how the grouping and blank lines worked for exports.

A new idea I have is to ignore the `groups` option for exports and never try to group them automatically. Instead, for exports you’ll group them manually by using comments.

```js
export * from "some-package"
export * from "./utils"
export * from "./Node"
export * from "./Edge"
```

⬇️ sort

```js
export * from "./Edge"
export * from "./Node"
export * from "./utils"
export * from "some-package"
```

⬇️ edit

```diff
 export * from "./Edge"
 export * from "./Node"
+export * from "some-package"
+// Utils
 export * from "./utils"
+export * from "./graph-utils"
-export * from "some-package"
```

⬇️ sort

```js
export * from "./Edge"
export * from "./Node"
export * from "some-package"

// Utils
export * from "./graph-utils"
export * from "./utils"
```

⬇️ edit and sort

```js
// Everything from some-package is exposed because blah blah
export * from "some-package"

// Components
export * from "./Edge"
export * from "./Node"

// Utils
export * from "./graph-utils"
export * from "./utils"
```

The idea is:

You choose what is grouped via comments. The groups themselves don’t change order. But inside each group everything is sorted alphabetically.

- What do you think about the “group via comments” idea?
- Can you try #53 on your projects and see what it looks like?
- Can you share some good examples of re-exports you want sorted from your projects?

**Comment by remcohaszing** [comment]
Actually my use case is the simplest one possible. My packages contain an `index.ts` which re-exports named packages.

Examples:

- https://gitlab.com/appsemble/appsemble/-/blob/master/packages/node-utils/src/index.ts
- https://gitlab.com/appsemble/appsemble/-/blob/master/packages/preact-components/src/index.ts
- https://gitlab.com/appsemble/appsemble/-/blob/master/packages/react-components/src/index.ts
- https://gitlab.com/appsemble/appsemble/-/blob/master/packages/utils/src/index.ts
- https://gitlab.com/appsemble/appsemble/-/blob/master/packages/web-utils/src/index.ts

Also I have moved away from this plugin, because the `import/order` ESLint rule can do pretty much the same now, works with `require`, and doesn’t require me to install a separate package. I may want to use this plugin again if re-exports are supported though.

**Comment by remcohaszing** [comment]
I do think the _“group via comments”_ idea is neat. I don’t need it for my projects (so far), but I do see its value for other projects.

**Comment by lydell** [comment]
Released in v6.0.0.
